scintillometry/psrio/psrfits_io.py

```diff
@@ -0,0 +1,383 @@
+    def frequency(self, frequency):
+        #NOTE this assumes the frequency marker is on the center of the channels
+        if hasattr(frequency, 'unit'):
+            frequency = frequency.to(u.MHz)
+        else:
+            frequency = u.Quantity(frequency, u.MHz, copy=False)
+        n_chan = len(frequency)
+        self.header['OBSNCHAN'] = n_chan
+        self.header['OBSBW'] = (frequency[-1] - frequency[0]).value
+        self.header['OBSFREQ'] = frequency[int(n_chan / 2)].value
+
+    @property
+    def ra(self):
+        return Longitude(self.header['RA'], unit=u.hourangle)
+
+    @ra.setter
+    def ra(self, val):
+        val = Longitude(val, unit=u.hourangle)
+        self.header['RA'] = val.to_string(sep=':')
+
+    @property
+    def dec(self):
+        return Latitude(self.header['DEC'], unit=u.deg)
+
+    @ra.setter
+    def dec(self, val):
+        val = Latitude(val, unit=u.deg)
+        self.header['DEC'] = val.to_string(sep=':')
+
+    @property
+    def obs_mode(self):
+        return self.header['OBS_MODE']
+
+
+class SubintHDU(fits.BinTableHDU, HDUBase):
+    """SubintHDU class provides the translator functions between baseband-style
+    file object and the PSRFITS SUBINT HDU.
+
+    Parameter
+    ---------
+    header_hdu : PsrfitsHearderHDU
+        The psrfits main header object
+    subint_hdu : HDU oject
+        The psrfits data HDU.
+
+    Note
+    ----
+    Right now we are assuming the data rows are continuous in time and the
+    frequency are the same.
+    """
+
+    _properties = ('start_time', 'sample_rate', 'shape', 'samples_per_frame',
+                   'polarization', 'frequency')
+    _header_defaults = subint_header
+    _req_columns = subint_columns
+
+    def __init__(self, header_hdu, subint_hdu=None):
+        self.header_hdu = header_hdu
+        if subint_hdu is None:
+            super(SubintHDU, self).__init__(name='SUBINT')
+            self._init_empty()
+        else:
+            super(SubintHDU, self).__init__(header=subint_hdu.header,
+                                            data=subint_hdu.data)
+        self.verify()
+        self.offset = 0
+
+    def _init_empty(self):
+        for k in self._header_defaults:
+            self.header.set(k[0], k[1], k[2])
+
+    def make_columns(self, name):
+        cols = []
+        for cl in self._columns_defaults:
+            col = fits.Column()
+
+    def verify(self):
+        if self.header['EXTNAME'].replace(' ', '') != "SUBINT":
+            raise ValueError("Input HDU is not a SUBINT type.")
+
+    @property
+    def mode(self):
+        return self.header_hdu.obs_mode
+
+    @property
+    def sample_rate(self):
+        # We assume the sample rate are the same for all the rows.
+        sample_time = u.Quantity(self.data[0]['TSUBINT'] /
+                                 self.samples_per_frame, u.s)
+        return 1.0 / sample_time
+
+    @sample_rate.setter
+    def sample_rate(self, val):
+        if hasattr(val, 'unit'):
```

Comment:
This is not OK, `val` is assumed to be in units of second when no unit is given. Let's just assume a `Quantity`, and do `self.header['TBIN'] = (1 / val).to_value(u.s)`